### ISSUE TITLE
Translating root_page.slug breaks child_page.url when changing language

### ISSUE THREAD

**Issue by dwasyl** [infoportugal/wagtail-modeltranslation#195]
See https://github.com/infoportugal/wagtail-modeltranslation/issues/195#issuecomment-376732049 for a detailed explanation of the issue and see https://github.com/dmarcelino/wagtail-modeltranslation/commit/5648c0b16412cb54cb7b780824b63d410245d8b9 for a failing test.

For now the workaround is to not translate a root page slug.

Edited by @dmarcelino

---

Hi all,

This is another new issue that didn't exist in 0.6, but when accessing a `Page` objects `url` property, translation seems to fail.

For example, in a shell:
```
from django.utils import translation
translation.activate('en')
for p in Page.objects.all():
    p.url
    p.slug

translation.activate('fr')
for p in Page.objects.all():
    p.url
    p.slug
```

In the above example, the `url` property will print just fine on the first run, but once the language changes `url` property is blank - although the `slug` property returns the correct value based on the activated language.

This wasn't an issue in 0.6.

Any thoughts?

EDIT: This seems to also cause `change_lang` to return None (similar to Issue #13)

**Comment by dmarcelino** [comment]
Hi @dwasyl,

From looking at what `Page.get_url()` does ([models.py#L744-L771](https://github.com/wagtail/wagtail/blob/master/wagtail/core/models.py#L744-L771)) it looks like you don't have `url_path` properly populated on the `wagtailcore_page` table (`Page` model). I'm guessing something went wrong when you upgraded and those didn't get set. Does this happen with pages you translated after migrating to 0.8? You can try running [set_translation_url_paths](http://wagtail-modeltranslation.readthedocs.io/en/latest/management%20commands.html#the-set-translation-url-paths-command) which will set `url_path` for all pages and languages.

Note that if you are using `Wagtail>=1.11`, you should replace `page.url` with `page.get_url(request=request)`, see http://docs.wagtail.io/en/v2.0/releases/1.11.html#upgrade-considerations.

**Comment by dwasyl** [comment]
Hi @dmarcelino,

From checking in `SQL`, the `url_path_XX` values are properly populated, although I've run `set_translation_url_paths` again as well.

As well, when I start a new shell and run:
```
translation.activate('fr')
for p in Page.objects.all():
    p.url
    p.slug
```
Everything is fine, The same is fine if I activate english `en` to start, the only time it has problems is when the activated language changes.

I've also noticed `routablepageurl` fails when the site is on the non-primary language, specifically, `base_url = page.relative_url(request.site)` returns None ([wagtailroutablepage_tags.py#L21](https://github.com/wagtail/wagtail/blob/master/wagtail/contrib/routable_page/templatetags/wagtailroutablepage_tags.py#L21)), this causes `routablepageurl` to throw a `TypeError`.

From some additional testing with `change_lang`, it also returns None, based on it's use of `.url` ([wagtail_modeltranslation.py#L44-L51](https://github.com/infoportugal/wagtail-modeltranslation/blob/master/wagtail_modeltranslation/templatetags/wagtail_modeltranslation.py#L44-L51)). That said, wouldn't using `page.get_url(request=request)` open `modeltranslation` up to returning a cached value for `.url`?

Few more testing results:
-  Manually entered fr paths load fine
- On english and french pages, `change_lang` and other methods return the correct 'en' url, but not the `fr` on either.
- Newly created pages have the same issue as migrated pages

**Comment by dmarcelino** [comment]
Hey @dwasyl,

Regarding the shell example, `translation.activate('fr')` is unlikely to work outside a view (see django-modeltranlations's [Accessing Translated Fields Outside Views](http://django-modeltranslation.readthedocs.io/en/latest/caveats.html?highlight=activate#accessing-translated-fields-outside-views) for details). You should try using `django.utils.translation.trans_real.activate('fr')`.

I've never used the `change_lang` tag, I use Django's [set_language](https://docs.djangoproject.com/en/1.11/topics/i18n/translation/#django.views.i18n.set_language), so I can't help you much there. But I do use `page.get_url(request=request)` to build a tree navigation menu and so far I haven't encountered any issues, either with the native language or with the translated language.

Any chance you can write a failing test that exposes the issues you are experiencing? `test_get_url_parts` and `test_url` at [tests.py#L774-L834](https://github.com/infoportugal/wagtail-modeltranslation/blob/master/wagtail_modeltranslation/tests/tests.py#L774-L834) should be a good example.

**Comment by dwasyl** [comment]
Thanks @dmarcelino, I've done some more debugging and come up with a reason.

To start, way back when, I originally configured a translated `Page` model, but also the `Site` model. This was so I could display both language versions of the `site.name` since on my sites there was a need for an overall site name, not just on the root page. This worked fine with the old model of `wagtail-modeltranslation` as enough of the `Page` functions were overridden and it worked. With the new version it seems to have caused it to break.

The problem is that in [wagtail/core/models.py#L153-L167](https://github.com/wagtail/wagtail/blob/master/wagtail/core/models.py#L153-L167) the first line pulls up the cached copy of  `wagtail_site_root_paths`, this returns the previously generated `site` object and last `root_page.url_path`. When the `Site` model is translated, `site.root_page.url_path` displayed a translated slug for the `root` page.

For example, Site A with a root page of `my-site`, the `site.root_page.url_path` would be `/my-site/`, if Site A is included in `modeltranslation`, there could be a separate language slug i.e. `site.root_page.url_path_fr` that equals `/mon-site/`. This is a problem because `models.py` `get_url_parts` checks the `url_path` of a particular page against the Site object's `root_path`. If the `root_page` has multiple slugs and a cached version is referenced this won't part and `get_url_parts` returns None, causing a whole bunch of other things to fail.

Anyway, thanks for the tips on this, but the end result is to make sure all language versions of the `Site`'s root page have the same slug (which is never visible anyway).

**Comment by francois-travais** [comment]
I could arrive to the same conclusion than @dwasyl by using the template tag `change_lang`.
My site root page has a different slug for each language and in `change_lang`, `page.url` returns `None`. Same result if I change  `page.url` with `page.get_url(request=request)`.

As soon as I set the same slug on the site root page for all languages then everything works fine.

**Comment by dmarcelino** [comment]
Thanks @dwasyl, that's great work! I've managed to reproduce the issue on my project by translating the root_page slugs and also came up with a failing test for this issue, see: https://github.com/dmarcelino/wagtail-modeltranslation/commit/31e3fe849225518c0faeec6ff96f7ce5147da0c3

@dwasyl, @francois-travais, I don't have the time for looking into a fix now but I'll gladly review any PRs.

**Comment by dwasyl** [comment]
Great @dmarcelino! Glad it isn't just me. Was a strange bug to track down.

My concern with the patches for `_get_site_root_paths()` or `get_site_root_paths()` is that they will get hit quite often. For example, on a page I load between menus, in page links and then a button for another language, it gets hit three or four times. Monkey patching wagtail enough to make the cache language aware would be ideal, but seems quite patched together (this is basically what I did on 0.6 on Wagtail 1.6 when there was much more overriding going on).

From the other side, is there any opportunity on the `modeltranslation` side to force users to have a single `slug` on the root page? This would prevent any monkey patching and shouldn't affect any sites since the root page slug gets hidden from the end user.

**Comment by dmarcelino** [comment]
Hey @dwasyl,

I understand your apprehension in patching, I'm not a fan of it either, unfortunately Wagtail doesn't come with means for us to provide our own `Page` model. A swappable `Page` model (as proposed in wagtail/wagtail#3282) would allow for a much cleaner implementation of `wagtail-modeltranslation` but it doesn't look like that will become a reality anytime soon.

Currently there is not a good way of checking if a page is a root page. We would first need to get the `Page`'s `Site` to see if the `Page` is a root page. It also doesn't help that we create a `Page` first an only after add it to a `Site` as a root page. Another thing I'm unsure of is if a `Page` can simultaneously be a root page in a `Site` and a normal page on another `Site`. I think it's best if we tackle this issue at its root `Site.get_site_root_paths()`.

**Comment by agelinas** [comment]
I'm having the same problem when switching to another language (From French to English). Will have to wait for a fix before using wagtail-modeltranslation in production.

Otherwise, starting a new installation worked fine for me (minus the problem).

Thank you everyone!

**Comment by dmarcelino** [comment]
Hi @agelinas, the workaround for this issue is to not translate the root page slug. If you've already translated them, deleting the translation(s) should fix the issue. This does not have visible effects since the root page slug is hidden from users.

**Comment by dmarcelino** [comment]
Fix proposed on PR #197, I'd appreciate if the folks on this thread could test it.

@dwasyl, please review PR #197 and let me know your thoughts.

**Comment by agelinas** [comment]
Sorry @dwasyl, I tried your fix for PR #197, but it did not work for me. I did change the root_path slugs so both my En and FR slugs were the same, as you suggested, and this did fix the problem. Thank you for your help.

**Comment by dmarcelino** [comment]
@agelinas, I've found an issue with the previous approach and tackled the problem in a different way (https://github.com/infoportugal/wagtail-modeltranslation/pull/197#issuecomment-378664170), please test the latest version:

```
pip install -e git+https://github.com/dmarcelino/wagtail-modeltranslation.git@a92cf200a3880e95b70f2a6a0a785339a8e4ac27#egg=wagtail-modeltranslation
```

**Comment by agelinas** [comment]
Tried the fix (for #197) and it worked like a charm! 

Thank you!

**Comment by dwasyl** [comment]
@dmarcelino have been offline for a few days, but trying out this latest build things seem to be working for me as well now.

**Comment by dmarcelino** [comment]
@agelinas, @dwasyl good to hear that!
